[image] Omit `children` in the props type

ashaller2017 opened the pull request:
Omit children from ImageProps

# Why

Want to have type hinting when including children in Image props. Should use ImageBackground when needing children. 

# How

Update the ImageProps type to omit children.

# Test Plan

Created fresh expo project importing my image type. Type hints when trying to pass children and doesn't when self closing. 

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

ashaller2017 committed:
Update Image.types.ts

Omit children from ImageProps

Co-Authored-By: theBGuy <60308670+thebguy@users.noreply.github.com>

ashaller2017 committed:
Update CHANGELOG.md

Co-Authored-By: theBGuy <60308670+thebguy@users.noreply.github.com>

ashaller2017 committed:
Update packages/expo-image/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

tsapeta commented on packages/expo-image/CHANGELOG.md:
```diff
@@ -10,6 +10,12 @@
 
 ### 💡 Others
 
+## 2.0.3 — 2024-11-23
+
+- Update `ImageProps` so `children` are omitted. ([#33afe30](https://github.com/expo/expo/pull/33210) by [@ashaller2017](https://github.com/ashaller2017)) ([#33210](https://github.com/expo/expo/pull/33210) by [@ashaller2017](https://github.com/ashaller2017))
+
+### 💡 Others
+
```
Changelog entries should be put under "Unpublished" section. Version and the date is automatically added by the publish script.
```suggestion
- Update `ImageProps` type so `children` are omitted. ([#33210](https://github.com/expo/expo/pull/33210) by [@ashaller2017](https://github.com/ashaller2017))

```

tsapeta commented on packages/expo-image/src/Image.types.ts:
```diff
@@ -89,7 +89,7 @@ export type ImageDecodeFormat = 'argb' | 'rgb';
  * but all of them are deprecated and might be removed in the future.
  */
-export interface ImageProps extends Omit<ViewProps, 'style'> {
+export interface ImageProps extends Omit<ViewProps, 'style' | 'children'> {
```
As this changes types, you also need to run `yarn build` in `packages/expo-image` to rebuild TS definitions in the build folder

ashaller2017 replied:
Ok on it! Thank you.

tsapeta reviewed (changes requested):
Thanks! Left two comments below.

ashaller2017 committed:
Update CHANGELOG.md

Update change log

Co-Authored-By: theBGuy <60308670+thebguy@users.noreply.github.com>

ashaller2017 committed:
Merge branch '@arishaller2017/image/omit-children-from-image-type' of https://github.com/ashaller2017/expo into @ashaller2017/image/omit-children-from-image-type

ashaller2017 committed:
Update packages/expo-image/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

ashaller2017 committed:
Remove logger and update expo-image build

Remove the warning logger since type hinting is now implemented.

Run yarn to rebuild expo-image types in build folder.

Co-Authored-By: theBGuy <60308670+thebguy@users.noreply.github.com>

ashaller2017 committed:
Merge branch '@arishaller2017/image/omit-children-from-image-type' of https://github.com/ashaller2017/expo into @arishaller2017/image/omit-children-from-image-type

ashaller2017 committed:
Update CHANGELOG.md

Followed @tsapeta suggestion.

tsapeta commented on packages/expo-image/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Unpublished
 
+- Update `ImageProps` type so `children` are omitted. ([#33210](https://github.com/expo/expo/pull/33210) by [@ashaller2017](https://github.com/ashaller2017))
```
It's still in a wrong place. Can you please move it to "Others" category?

ashaller2017 replied:
Read previous comment wrong. Apologies.

tsapeta commented on packages/expo-image/src/Image.tsx:
```diff
@@ -199,13 +199,6 @@ export class Image extends React.PureComponent<ImageProps> {
-        'The <Image> component does not support children. If you want to render content on top of the image, consider using the <ImageBackground> component or absolute positioning.'
-      );
-      loggedRenderingChildrenWarning = true;
-    }
```
I think we don't want to remove that warning – it's good to let the user know if he is doing something wrong

ashaller2017 replied:
When building Expo-Image, Omitting the children will cause a build failure. This initial logger was setup due to a discord discussion we had when it wasn't clear you could not wrap children in an image component. By type hinting this error should be apparent. I could rebuild throwing a @ts-ignore to build this. What do you think?

tsapeta replied:
In development Metro bundler will not complain if you have issues in types, it strips out all the types. Your code editor will hint you about the mismatch in types, but I think this is not enough. Especially when your app is big, you don't immediately see all such issues when you upgrade the package until the component is used.

ashaller2017 committed:
Update Image.tsx

Add expect error to restProps.children from new type.

ashaller2017 committed:
Update CHANGELOG.md

Move comment to Others

ashaller2017 committed:
Update Image.d.ts.map

Rebuild after adding logger again.

tsapeta commented on packages/expo-image/src/Image.tsx:
```diff
@@ -198,14 +198,14 @@ export class Image extends React.PureComponent<ImageProps> {
       loggedRenderingChildrenWarning = true;
     }
-
+    
```
Linter is failing on this 😉 Don't forget to rebuild files after this is applied (not sure if this is needed for just whitespace changes, probably not).
```suggestion

```

ashaller2017 committed:
Remove whitespace and rebuild.

ashaller2017 committed:
Merge branch 'main' into @arishaller2017/image/omit-children-from-image-type

ashaller2017 committed:
Merge branch 'main' into @arishaller2017/image/omit-children-from-image-type

ashaller2017 committed:
Update Image.types.d.ts.map

Rebuild image after merge from main.
